File: smart_product_ai_frontend/src/api/socket.ts
import { BACKEND_ORIGIN } from "./config";

/*
    Real time AI progress.
    Instead of refreshing:
    Processing...
    20%
    80%
    Completed
    arrives instantly.
*/
export function connectAnalysisSocket(
analysisId:string,
callback:(data:any)=>void
){
const websocketOrigin = BACKEND_ORIGIN
    ? BACKEND_ORIGIN.replace(/^http/, "ws")
    : `${window.location.protocol === "https:" ? "wss:" : "ws:"}//${window.location.host}`;
const socket =
new WebSocket(
`${websocketOrigin}/api/v1/ws/analysis/${analysisId}`
);
socket.onmessage =
(event)=>{
    const data =
    JSON.parse(
        event.data
    );
    callback(data);
};

return socket;

}

File: app/api/v1/socket.py
from fastapi import APIRouter, WebSocket

router = APIRouter()

@router.websocket(
    "/ws/analysis/{analysis_id}"
)
async def analysis_socket(
    websocket: WebSocket,
    analysis_id:str
):
    """
    Real-time AI progress.
    Frontend receives:
    queued
    processing
    completed
    """
    await websocket.accept()
    await websocket.send_json(
        {
        "analysis_id": analysis_id,
        "status": "processing"
        }
    )
    await websocket.close()